osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,237 @@
+#include <osquery/utils/conversions/join.h>
+#include <osquery/utils/conversions/split.h>
+#include <osquery/utils/conversions/windows/strings.h>
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+const std::string kEventLogXmlPrefix = "<QueryList><Query Id=\"0\">";
+const std::string kEventLogXmlSuffix = "</Query></QueryList>";
+
+const int kNumEventsBlock = 1024;
+
+void parseWelXml(QueryContext& context,
+                 std::wstring& xml_event,
+                 RowYield& yield) {
+  auto row = make_table_row();
+  pt::ptree propTree;
+  WELEvent windows_event;
+  auto status = parseWindowsEventLogXML(propTree, xml_event);
+  status = parseWindowsEventLogPTree(windows_event, propTree);
+
+  row["time"] = INTEGER(windows_event.osquery_time);
+  row["datetime"] = SQL_TEXT(windows_event.datetime);
+  row["channel"] = SQL_TEXT(windows_event.source);
+  row["provider_name"] = SQL_TEXT(windows_event.provider_name);
+  row["provider_guid"] = SQL_TEXT(windows_event.provider_guid);
+  row["eventid"] = INTEGER(windows_event.event_id);
+  row["task"] = INTEGER(windows_event.task_id);
+  row["level"] = INTEGER(windows_event.level);
+  row["pid"] = INTEGER(windows_event.pid);
+  row["tid"] = INTEGER(windows_event.tid);
+
+  row["keywords"] = SQL_TEXT(windows_event.keywords);
+  row["data"] = SQL_TEXT(windows_event.data);
+
+  if (context.hasConstraint("time_range", EQUALS)) {
+    auto time_range = context.constraints["time_range"].getAll(EQUALS);
+    row["time_range"] = SQL_TEXT(*time_range.begin());
+  } else {
+    row["time_range"] = SQL_TEXT("");
+  }
+
+  if (context.hasConstraint("timestamp", EQUALS)) {
+    auto timestamp = context.constraints["timestamp"].getAll(EQUALS);
+    row["timestamp"] = SQL_TEXT(*timestamp.begin());
+  }
+
+  if (context.hasConstraint("xpath", EQUALS)) {
+    auto xpaths = context.constraints["xpath"].getAll(EQUALS);
+    row["xpath"] = SQL_TEXT(*xpaths.begin());
+  }
+
+  yield(std::move(row));
+}
+
+void parseQueryResults(QueryContext& context,
+                       EVT_HANDLE queryResults,
+                       RowYield& yield) {
+  std::vector<EVT_HANDLE> events(kNumEventsBlock);
+  unsigned long numEvents = 0;
+
+  // Retrieve the events one block at a time
+  auto ret = EvtNext(
+      queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+
+  while (ret != FALSE) {
+    for (unsigned long i = 0; i < numEvents; i++) {
+      unsigned long renderedBuffSize = 0;
+      unsigned long renderedBuffUsed = 0;
+      unsigned long propCount = 0;
+      if (!EvtRender(nullptr,
+                     events[i],
+                     EvtRenderEventXml,
+                     renderedBuffSize,
+                     nullptr,
+                     &renderedBuffUsed,
+                     &propCount)) {
+        if (GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
+          LOG(WARNING) << "Failed to allocated memory to render event "
```

Comment:
Isn't this error actually something like "Failed to get size of rendered event"?